### DISCUSSION TITLE
Tabs vs. Spaces?

### DISCUSSION THREAD

**Discussion by tziemek** [artiso-solutions/adsd-hotel#33] [General] [answered]
What's your preferred way? We should commit to one style!

**Comment by tommasobertoni** [comment]
Spaces.

Tabs look weird:
```
one	tab	of	separation
```

```
four    spaces    of    separation
```

and may change in different IDEs (vscode)
![image](https://user-images.githubusercontent.com/8939890/98228018-55ae4700-1f58-11eb-80f9-ae91c732c81f.png)

**Comment by tommasobertoni** [comment, marked as answer]
More generally, I'd go with the [aspnetcore guidelines](https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines#coding-style-guidelines--general):

1. Use four spaces of indentation (no tabs)
2. Use `_camelCase` for private fields
3. Avoid `this.` unless absolutely necessary
4. Always specify member visibility, even if it's the default (i.e. `private string _foo`; not `string _foo;`)
5. Open-braces (`{`) go on a new line
6. Use any language features available to you (expression-bodied members, throw expressions, tuples, etc.) as long as they make for readable, manageable code.
7. This is pretty bad: `public (int, string) GetData(string filter) => (Data.Status, Data.GetWithFilter(filter ?? throw new ArgumentNullException(nameof(filter))));`

Proposal: [add](https://docs.microsoft.com/en-us/visualstudio/ide/create-portable-custom-editor-options?view=vs-2019) an `.editorconfig` at the solution level.

**Reply by tziemek** [reply]
I think we can enforce, or at least suggest, 1-5 with an editorconfig file. I pushed mine to the repo in #13  and we can tweak that. Personally I don't like the `_` prefix for fields, but that's open for discussion.

**Reply by tziemek** [reply]
Any opinions? @antonio8101  @exbachi  @dpetscher  @tobias-tengler  @FischlerA

**Reply by exbachi** [reply]
> I think we can enforce, or at least suggest, 1-5 with an editorconfig file. I pushed mine to the repo in #13 and we can tweak that. Personally I don't like the `_` prefix for fields, but that's open for discussion.

i dont like the underscore prefix. in my opinion, we can leave it out.

**Reply by exbachi** [reply]
I like tabs more, but I really don't care that much, if we use more than two whitespaces.

**Reply by antonio8101** [reply]
I would suggest to use the _ as prefix for all the private fields. I find it auto well self explanatory, and expect not to use the this keyword instead.

**Reply by tobias-tengler** [reply]
I'm also for the underscore and all the other conventions Tommaso has laid out!

**Reply by tziemek** [reply]
You are only used to the R# defaults^^.

**Reply by tommasobertoni** [reply]
Proposal:
- do enforce 4 spaces for source files indentation
- do enforce 2 spaces for xml/yaml files indentation
- do _not_ enforce underscore prefix on private fields: let each team decide
(this has the most indecision)

**Reply by tziemek** [reply]
The first is already in the editorconfig file. Second one only for xml, I'll add yaml.
But we should reach an agreement on the underscore for private fields. I don't like it, but I can go with it. VS offers refactorings and R# does it by default, so I have no problem.

**Reply by exbachi** [reply]
If the majority supports underscore, I'm fine with it.

**Reply by tziemek** [reply]
I've put that in the editorconfig, already in `main`. I close the discussion now.
